File: src/main/java/tw/racket/lang/RacketParserDefinition.java
package tw.racket.lang;

import com.intellij.lang.ASTNode;
import com.intellij.lang.ParserDefinition;
import com.intellij.lang.PsiParser;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.project.Project;
import com.intellij.psi.FileViewProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IFileElementType;
import com.intellij.psi.tree.TokenSet;
import org.jetbrains.annotations.NotNull;
import tw.racket.lang.lexer.RacketLexerAdapter;
import tw.racket.lang.parser.RacketParser;
import tw.racket.lang.psi.RacketFile;
import tw.racket.lang.psi.RacketTypes;

public class RacketParserDefinition implements ParserDefinition {
    public static final TokenSet WHITE_SPACES = TokenSet.create(TokenType.WHITE_SPACE);
    public static final TokenSet COMMENTS = TokenSet.create(RacketTypes.COMMENT);

    public static final IFileElementType FILE = new IFileElementType(RacketLanguage.INSTANCE);

    @NotNull
    @Override
    public Lexer createLexer(Project project) {
        return new RacketLexerAdapter();
    }

    @Override
    public PsiParser createParser(Project project) {
        return new RacketParser();
    }

    @Override
    public IFileElementType getFileNodeType() {
        return null;
    }

    @NotNull
    @Override
    public TokenSet getWhitespaceTokens() {
        return WHITE_SPACES;
    }

    @NotNull
    @Override
    public TokenSet getCommentTokens() {
        return COMMENTS;
    }

    @NotNull
    @Override
    public TokenSet getStringLiteralElements() {
        return TokenSet.EMPTY;
    }

    @NotNull
    @Override
    public PsiElement createElement(ASTNode astNode) {
        return RacketTypes.Factory.createElement(astNode);
    }

    @Override
    public PsiFile createFile(FileViewProvider fileViewProvider) {
        return new RacketFile(fileViewProvider);
    }
}

File: src/main/kotlin/tw/racket/lang/RacketParserDefinition.kt
package tw.racket.lang

import com.intellij.lang.ASTNode
import com.intellij.lang.ParserDefinition
import com.intellij.lang.PsiParser
import com.intellij.lexer.Lexer
import com.intellij.openapi.project.Project
import com.intellij.psi.FileViewProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.TokenType
import com.intellij.psi.tree.IFileElementType
import com.intellij.psi.tree.TokenSet
import tw.racket.lang.lexer.RacketLexerAdapter
import tw.racket.lang.parser.RacketParser
import tw.racket.lang.psi.RacketFile
import tw.racket.lang.psi.RacketTypes

class RacketParserDefinition : ParserDefinition {
    override fun createLexer(project: Project): Lexer {
        return RacketLexerAdapter()
    }

    override fun createParser(project: Project): PsiParser {
        return RacketParser()
    }

    override fun getFileNodeType(): IFileElementType {
        return FILE
    }

    override fun getWhitespaceTokens(): TokenSet {
        return WHITE_SPACES
    }

    override fun getCommentTokens(): TokenSet {
        return COMMENTS
    }

    override fun getStringLiteralElements(): TokenSet {
        return TokenSet.EMPTY
    }

    override fun createElement(astNode: ASTNode): PsiElement {
        return RacketTypes.Factory.createElement(astNode)
    }

    override fun createFile(fileViewProvider: FileViewProvider): PsiFile {
        return RacketFile(fileViewProvider)
    }

    companion object {
        val WHITE_SPACES = TokenSet.create(TokenType.WHITE_SPACE)
        val COMMENTS = TokenSet.create(RacketTypes.COMMENT)
        val FILE = IFileElementType(RacketLanguage)
    }
}